### PR TITLE
Micro-benchmarks for matmul and HBM bandwidth

### PR DESCRIPTION
**Pull request by chishuen**
Add micro-benchmarks for matmul and HBM bandwidth tests.

TODO: move the code to https://github.com/AI-Hypercomputer/accelerator-microbenchmarks.git and add a shell script here pointing to that repo.

### CODE REVIEW

File: microbenchmarks/requirements.txt [hunk 1]
```diff
@@ -0,0 +1,2 @@
+jax[tpu]
+tensorboard-plugin-profile
```
Comment (bvandermoon) [inline microbenchmarks/requirements.txt:2]: Is this dependency needed? I think the profiling logic you are using is already available in JAX. Let me know if I am missing something

File: microbenchmarks/benchmark_matmul.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+      lambda: jax.block_until_ready(compiled(a, b)),
+      num_iter=args.num_iter,
+      warmup_iter=args.warmup_iter,
+      log_dir=args.log_dir,
+      func_label=args.label,
+      event_matcher=matcher,
+  )
+
+  # MXU is done in units of bf16.
+  mxu_bytes_per_op = 2
```
Comment (bvandermoon) [inline microbenchmarks/benchmark_matmul.py:130]: I think this is just for my understanding - why are we assuming bf16 here if we are supporting multiple dtypes?

File: microbenchmarks/requirements.txt (hunk 1 above)
Comment (chishuen) [reply microbenchmarks/requirements.txt]: Ah yes. `tensorboard-plugin-profile` removed. Thank you for catching that.

File: microbenchmarks/benchmark_matmul.py (hunk 2 above)
Comment (chishuen) [reply microbenchmarks/benchmark_matmul.py:130]: Good question. My original thinking was that this factor of 2 comes from the fact that *all multiplies (of a MXU) take bfloat16 inputs*: http://screen/3GDvMyTLntetKSz. But on a second thought, this seems unnecessary in the computation of FLOPs and can be confusing for data types other than bf16 like you mentioned. I'm removing this line and updating the calculation of FLOPs. PTAL.

Hi @Obliviour any thoughts?

File: microbenchmarks/benchmark_matmul.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+  result = run_bench(
+      lambda: jax.block_until_ready(compiled(a, b)),
+      num_iter=args.num_iter,
+      warmup_iter=args.warmup_iter,
+      log_dir=args.log_dir,
+      func_label=args.label,
+      event_matcher=matcher,
+  )
+
+  # 2 ops (multiple and add)
```
Comment (bvandermoon) [inline microbenchmarks/benchmark_matmul.py:129]: Nit: multiply

**Review by bvandermoon — APPROVED** [review]
LGTM pending the input from @Obliviour on the final calculations

File: microbenchmarks/benchmark_matmul.py (hunk 3 above)
Comment (chishuen) [reply microbenchmarks/benchmark_matmul.py]: Fixed. Thanks!

File: microbenchmarks/README.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+See https://cloud.google.com/tpu/docs/regions-zones for available zones.
+
+SSH into the VM:
+```
+gcloud compute ssh $TPU_NAME --zone=$ZONE
+```
+
+Clone the repo and install the dependencies:
+```bash
+git clone https://github.com/chishuen/tpu-recipes.git
```
Comment (Obliviour) [inline microbenchmarks/README.md:21]: nit on changing from your fork to the main repo

File: microbenchmarks/benchmark_hbm.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+          "A label used to name the function to be benchmarked in the trace"
+          " events."
+      ),
+  )
+  parser.add_argument(
+      "--matcher",
+      type=str,
+      required=False,
+      help=(
+          "A regex-based string matcher to filter the trace evnets eligible for"
```
Comment (Obliviour) [inline microbenchmarks/benchmark_hbm.py:96]: nit events

File: microbenchmarks/benchmark_matmul.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+      default=100,
+      help="Number of times the matmul kernel will be run.",
+  )
+  parser.add_argument(
+      "--warmup_iter",
+      type=int,
+      default="1",
+      help=(
+          "Number of times the matmul kernel will be run to warm up before the"
+          " acutal timing measurement starts."
```
Comment (Obliviour) [inline microbenchmarks/benchmark_matmul.py:77]: nit actual

File: microbenchmarks/benchmark_matmul.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+          "A label used to name the function to be benchmarked in the trace"
+          " events."
+      ),
+  )
+  parser.add_argument(
+      "--matcher",
+      type=str,
+      required=False,
+      help=(
+          "A regex-based string matcher to filter the trace evnets eligible for"
```
Comment (Obliviour) [inline microbenchmarks/benchmark_matmul.py:100]: nit events

File: microbenchmarks/benchmark_matmul.py [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+      lambda: jax.block_until_ready(compiled(a, b)),
+      num_iter=args.num_iter,
+      warmup_iter=args.warmup_iter,
+      log_dir=args.log_dir,
+      func_label=args.label,
+      event_matcher=matcher,
+  )
+
+  # 2 ops (multiply and add)
+  compute_flops = m * n * k * 2
```
Comment (Obliviour) [inline microbenchmarks/benchmark_matmul.py:130]: Just for my understanding to confirm:

since fp8 is not supported on trillium, we don't have another multiple of 2 here for fp8 case?

File: microbenchmarks/benchmark_utils.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+  """Filter the trace events eligible for benchmarking.
+
+  Args:
+    trace: a trace object in JSON format.
+    event_matcher: a regex-based event name matcher to filter the evnets.
+
+  Returns:
+    A list of events objects in JSON format.
+  """
+  if "traceEvents" not in trace:
```
Comment (Obliviour) [inline microbenchmarks/benchmark_utils.py:72]: I really like this trace event method of timing, one aside is I hope they don't change these names over time accidently breaking our test

**Review by Obliviour — APPROVED** [review]
LGTM, some generally small nits + 1 confirmation to understand the matmul tflop calucation. Thanks for the fast and great work @chishuen

File: microbenchmarks/README.md (hunk 4 above)
Comment (chishuen) [reply microbenchmarks/README.md:21]: Good catch 😆 fixed

File: microbenchmarks/benchmark_hbm.py (hunk 5 above)
Comment (chishuen) [reply microbenchmarks/benchmark_hbm.py:96]: done

File: microbenchmarks/benchmark_utils.py (hunk 9 above)
Comment (bvandermoon) [reply microbenchmarks/benchmark_utils.py:72]: +1 I don't think this method is endorsed by the profiler team

File: microbenchmarks/benchmark_matmul.py (hunk 6 above)
Comment (chishuen) [reply microbenchmarks/benchmark_matmul.py]: done

File: microbenchmarks/benchmark_matmul.py (hunk 7 above)
Comment (chishuen) [reply microbenchmarks/benchmark_matmul.py]: done

File: microbenchmarks/benchmark_matmul.py (hunk 8 above)
Comment (chishuen) [reply microbenchmarks/benchmark_matmul.py]: Here's my understanding of how the throughput is calculated. Please correct me if I'm wrong:

i. The number of required floating-point operations is independent of the data type.
ii. If the precision is reduced from bf16 to fp8, the completion time should be halved in theory, which results in doubled FLOPS. 
iii. Since fp8 is not supported on trillium (or not yet??), we expect similar FLOPS for bf16 vs fp8.

Does this makes sense?

File: microbenchmarks/benchmark_utils.py (hunk 9 above)
Comment (chishuen) [reply microbenchmarks/benchmark_utils.py:72]: That's a good point. I've been exploring different ways to measure the timing. Here are a couple of findings:

i. fhzhang@ and vipannalla@ recommend measuring timing from trace when it comes to measuring compute utilization (see [docs](http://shortn/_yb52aQDgtF)).
ii. xprof appears to have an internal API to get trace data in a less brittle way ([example](http://google3/platforms/xla/tools/misura/benchmark_utils.py?rcl=736001499&l=169-180)). But I'm not sure if they are available for external users.
iii. This pattern (e.g. using hard-coded strings like `traceEvents`) is not unprecedented internally ([example](http://google3/platforms/xla/mosaic/examples/benchmark.py?rcl=736001499&l=66-92)).

Given that this is an urgent ask, I'll stick with this approach for now. I'm working with qinyiyan@ and chandrasekhard@ to consolidate the microbenchmarks as a longer term solution. I'll see if we can find a more robust approach in that effort.
